include: zephyr: remove __ZEPHYR__

gmarull opened the pull request:
CMake always injects `__ZEPHYR__` to any code being built on Zephyr (e.g.
a module, an application, etc.). Requiring an include that defines
`__ZEPHYR__` to determine if I'm on Zephyr is problematic as well: it
requires including a Zephyr header (which will exist if building for...
Zephyr! unless a "mock" is provided).

Note that this change leaves <zephyr/zephyr.h> in a questionable
position: why shouldn't I just include <zephyr/kernel.h>?

gmarull committed:
include: zephyr: remove __ZEPHYR__

CMake always injects __ZEPHYR__ to any code being built on Zephyr (e.g.
a module, an application, etc.). Requiring an include that defines
__ZEPHYR__ to determine if I'm on Zephyr is problematic as well: it
requires to include a Zephyr header (which will exist if building for...
Zephyr!).

Note that this change leaves <zephyr/zephyr.h> in a questionable
position: why shouldn't I just include <zephyr/kernel.h>?

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

nashif reviewed (changes requested):
This will break a whole lot of out of tree code depending on this, I am not sure what exactly is the problem with this.

stephanosio commented on include/zephyr/zephyr.h:
```diff
@@ -7,14 +7,6 @@
 #ifndef ZEPHYR_INCLUDE_ZEPHYR_H_
 #define ZEPHYR_INCLUDE_ZEPHYR_H_
 
-/*
- * Applications can identify whether they are built for Zephyr by
- * macro below. (It may be already defined by a makefile or toolchain.)
- */
-#ifndef __ZEPHYR__
-#define __ZEPHYR__
-#endif
```
The rationale for this is given in the commit https://github.com/zephyrproject-rtos/zephyr/commit/0c6df2d55cd934c20f9add4f3fda7d59a1fc4051 from almost 5 years ago:

> Currently it's defined by Makefile, so if an app uses it's own build system, it won't be
available.

This is no longer valid (an app must use the Zephyr's build system to maintain everyone's sanity), and `zephyr.h` is not a "must-include" header either, so it only makes sense to remove this now.

stephanosio commented:
> This will break a whole lot of out of tree code depending on this

well, do they have their own independent build system (i.e. not an extension of the Zephyr's build system)?

gmarull commented:
> > This will break a whole lot of out of tree code depending on this
> 
> well, do they have their own independent build system (i.e. not an extension of the Zephyr's build system)? because AFAIU, it does not make sense to do that at this point.

Even if the build system is independent of Zephyr, it should be responsible to define `__ZEPHYR__` when building for Zephyr. Requiring a Zephyr include to determine if I'm on Zephyr means you already know beforehand you're on Zephyr... Because if not, you'll need to provide a mock include, basically.

carlescufi commented:
I agree with the change, but would like @tejlmand's opinion.

tejlmand reviewed (dismissed):
I think this change makes sense.

Especially as @stephanosio points out, `zephyr.h` is not a must-include.
Let's rely on the build system to set this (as is already the case).
And if an oot project has a very special way of integrating to Zephyr that avoids running Zephyr CMake code (sigh!) , then they should (must) be able to handle this themselves.

Thanks for cleaning this up.

nashif reviewed (changes requested):
What exactly is the cost of leaving this, what are we gaining removing this? It looks like the only reason for this is being removed is to free up zephyr.h and and put it on the chopping block next.

If we officially at a point where we are saying, you CAN'T build zephyr with anything else but the zephyr build system, then we should document this somewhere and make it clear, explain that \_\_ZEPHYR\_\_ only comes via  the build system etc.

also, related to third party build systems, why do we have this in the tree? `cmake/makefile_exports/CMakeLists.txt`. Commit d3a8fd40b902e1f1b95be9812d8b71dc0e270a1e clearly says:

>  This emulates the 'outputexports' target that KBuild supported and is
    supported for the same reasons that KBuild supported it. Easier
    integration with third-party build systems.

If the ultimate goal is getting rid of zephyr.h, it would be great to see a proposal and a path for its removal and a discussion on how to make this less painful for users.

I am just not a big fan of  seeing interfaces being dropped or changed in PRs without having a full conext of where things are going and what the ultimate goal is.

nashif commented:
Also, adding @pfalcon who was the source for this to confirm if all this still applies and to get his feedback.

gmarull commented:
> What exactly is the cost of leaving this, what are we gaining removing this? 

Remove legacy code? Consistency? Could you provide an example that relies on this nowadays?

> It looks like the only reason for this is being removed is to free up zephyr.h and and put it on the chopping block next.
> 

This patch has nothing to do with the fact that zephyr.h is equivalent to kernel.h. This is a separate topic.

> 
> 
> If we officially at a point where we are saying, you CAN'T build zephyr with anything else but the zephyr build system, then we should document this somewhere and make it clear, explain that \_\_ZEPHYR\_\_ only comes via  the build system etc.
> 


> 
> 
> also, related to third party build systems, why do we have this in the tree? `cmake/makefile_exports/CMakeLists.txt`. Commit d3a8fd40b902e1f1b95be9812d8b71dc0e270a1e clearly says:
> 
> 
> 
> >  This emulates the 'outputexports' target that KBuild supported and is
> 
>     supported for the same reasons that KBuild supported it. Easier
> 
>     integration with third-party build systems.
> 
> 
> 
> If the ultimate goal is getting rid of zephyr.h, it would be great to see a proposal and a path for its removal and a discussion on how to make this less painful for users.

Its removal is trivial: just include kernel.h, and keep zephyr.h for a while, but again, this is not what this PR is touching.

> I am just not a big fan of  seeing interfaces being dropped or changed in PRs without having a full conext of where things are going and what the ultimate goal is. 
> 

What interface is being dropped here? This is just removing something that is no longer useful. And, as explained, this wasn't great from the very beginning, there are better ways to solve this.

tejlmand reviewed (dismissed):
lgtm

gmarull commented:
ping @pfalcon @nashif

nashif commented:
> ping @pfalcon @nashif

my -1 stands. if we are officially removing support for 3rd party build systems, it has to be done across the tree and all code related to 3rd build systems need to be removed. This has to be documented and communicated. We do not know what we do not know. 
There is ZERO cost leaving this be, no impact whatsoever on anything.

stephanosio commented:
> if we are officially removing support for 3rd party build systems

I do not see how this is "removing support for 3rd party build systems." It has never been required/stated/documented that users **must** include the `zephyr.h` in their application, so it would be incorrect for one to rely on `zephyr.h` to define `__ZEPHYR__` anyway, not to mention that would be unnecessary because it is already defined by the build system.

Moreover, third-party build systems, which does not really make sense at this point but if you ever choose to do it for whatever reason, would be required to define `__ZEPHYR__` globally anyway for the reason stated above. As far as I understand, supporting third-party build systems (as in building the Zephyr itself using something other than the Zephyr's own CMake-based build system) has never been one of our goals and does not really make sense at this point due to the inherent complexity of the Zephyr build process and the heavy reliance on the Zephyr-specific macros in the CMake files.

Also what has been said in https://github.com/zephyrproject-rtos/zephyr/pull/47268#discussion_r913075244.

carlescufi commented:
As I stated before, my vote goes for removing this. As @stephanosio said, this does not break compatibility with 3rd-party toolchains and it is completely unnecessary.

gmarull commented:
I think the problem is in the design itself. Let's say picolibc wants to have certain Zephyr-specific code. It will use `__ZEPHYR__` guard for that. Now, picolibc goes and says, “oh, I need to include zephyr.h for that!” It includes it and everything “works fine”. Now, a poor guy using FreeRTOS + picolibc gets a build error: “Could not find <zephyr/zephyr.h>”, and he says “what the hell is that, I'm not using Zephyr! FreeRTOS is better! Please take that out!”. So, the header approach is problematic. It is the job of the build system (Zephyr or external) to tell it's building for Zephyr, same as I'll get `__llvm__` if compiling code on LLVM.

nashif commented:
since we missed the dev-review, my comment from above again:

> If we officially at a point where we are saying, you CAN'T build zephyr with anything else but the zephyr build system, then we should document this somewhere and make it clear, explain that __ZEPHYR__ only comes via the build system etc.

This line in zephyr.h was added for a reason (supporting 3rd party build systems), while not directly documented, like many other things in zephyr, it is something that someone might depend on, could have depended on or something else we do not know. The fact is it is a line in a public header.

If we are are going to remove something supporting a "feature", documented or not, we have to review the whole thing and not just drop a flag to reduce the number of line in zephyr.h, clearly this removal is not fixing a direct issue.

I am not saying we should keep it and I am not going into the argument of why this is good or bad, what I am asking here is  clarify through documentation that to build zephyr, you have to use the zephyr build system, third party build systems are not support ("anymore) and that you should not depend on __ZEPHYR__ coming from headers, etc. etc.
While we are at it, do a sweep and see if other thing like the item below (exports) exist to support third part build systems. The exports feature is could still be useful, unrelated to 3rd party build systems.

This could be done in the release notes, in the build system section or wherever you want.

We should manage this type of cleanups with more care and always consider that someone out there might be using this  while we continue on improving public interfaces and documentation for end-users. 

> also, related to third party build systems, why do we have this in the tree? cmake/makefile_exports/CMakeLists.txt.

gmarull committed:
doc: release-notes: inform about __ZEPHYR__ removal from zephyr.h

zephyr.h header no longer defines __ZEPHYR__, add some details on the
release notes.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>
